posydon/unit_tests/utils/test_posydonwarning.py

```diff
@@ -0,0 +1,747 @@
+        totest.warnings.filterwarnings(action='always',\
+                                       category=ResourceWarning)
+        for i in range(10):
+            assert totest._apply_POSYDON_filter(warning={'message': "Test"+\
+                   str(i), 'category': ResourceWarning}) == {'message': \
+                   "Test"+str(i), 'category': ResourceWarning}
+            for k,v in totest._POSYDON_WARNINGS_REGISTRY.items():
+                if "ResourceWarning" in k:
+                    assert i == v
+        # check the route of an error filter
+        totest.warnings.filterwarnings(action='error',\
+                                       category=ResourceWarning)
+        assert totest._apply_POSYDON_filter(warning={'message': "Test"+str(i),\
+               'category': ResourceWarning}) == {'message': "Test"+str(i),\
+               'category': ResourceWarning}
+
+    def test_issue_warn(self, capsys, monkeypatch, recwarn, clear_registry):
+        def mock_apply_POSYDON_filter(warning=dict(message="No warning"), registry=None):
+            return None
+
+        # wrong arguments
+        with raises(TypeError, match="warning must be a dictionary"):
+            totest._issue_warn(warning="Test")
+        with raises(TypeError, match="message must be a string"):
+            totest._issue_warn(warning={'message': 1})
+        with raises(TypeError, match="stacklevel must be an integer"):
+            totest._issue_warn(warning={'stacklevel': "Test"})
+        with raises(TypeError, match="registry must be a dictionary or None"):
+            totest._issue_warn(registry="Test")
+        # check default warning
+        with warns(UserWarning, match="No warning") as warn1:
+            totest._issue_warn()
+        assert len(warn1) == 1
+        # check filtered warning
+        monkeypatch.setattr(totest, "_apply_POSYDON_filter", mock_apply_POSYDON_filter)
+        totest._issue_warn()
+        assert len(recwarn) == 0
+
+    def test_Pwarn(self):
+        # missing argument
+        with raises(TypeError, match="missing 1 required positional argument"):
+            totest.Pwarn()
+        # wrong arguments
+        with raises(TypeError, match="message must be a string"):
+            totest.Pwarn(1)
+        with raises(TypeError, match="stacklevel must be an integer"):
+            totest.Pwarn("Unit", stacklevel="Test")
+        # check output of POSYDONwarning
+        with warns(totest.POSYDONWarning, match="Unit test") as warn1:
+            totest.Pwarn("Unit test", "POSYDONWarning")
+        assert len(warn1) == 1
+        # check output of FutureWarning
+        with warns(FutureWarning, match="Unit test") as warn2:
+            totest.Pwarn("Unit test", FutureWarning)
+        assert len(warn2) == 1
+        # check output of UserWarning (default if unspecified)
+        with warns(UserWarning, match="Unit test") as warn3:
+            totest.Pwarn("Unit test")
+        assert len(warn3) == 1
```

Comment:
the `warn1`, `warn2`, and `warn3` are technically out of scope. They should be inside the `with ....` indentation.

Reply:
This is the recommended usage by pytest: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
I think it is even the case, that inside the block there is no guaranty that the object of this context manager is updated, it may update it properly only on exit.
I'm not 100% sure, how to interpret the scope here, because this object is created before the block. Thus, I'd interpret it more similar to a function object, created by `def`, which is not only valid inside the scope of the function.